### PR TITLE
ML-613: not writing value for empty key

### PR DESCRIPTION
**Pull request by katyakats**
https://jira.iguazeng.com/browse/ML-589

### CODE REVIEW

File: tests/test_flow.py [hunk 1]
```diff
@@ -2556,6 +2556,19 @@ def test_custom_string_time_input():
...
+    data = pd.DataFrame({'first_name': ['moshe', None, 'katya'], 'some_data': [1, 2, 3]})
+    data.set_index(keys=['first_name'], inplace=True)
+
+    controller = build_flow([
+        DataframeSource(data, key_field=['first_name']),
+        Reduce([], append_and_return),
+    ]).run()
+    result = controller.await_termination()
+
+    assert len(result) == 2
```
Comment (gtopper) [inline tests/test_flow.py]: I know the cardinality is the point, but still better to make sure we got the right two elements.

File: storey/sources.py [hunk 2]
```diff
@@ -619,20 +620,33 @@ def _blocking_io_loop(self):
...
                         if self._time_field:
                             time_field = self._time_field
                             if self._with_header and isinstance(time_field, str):
                                 time_field = field_name_to_index[time_field]
                             time_as_datetime = parsed_line[time_field]
                         else:
                             time_as_datetime = datetime.now()
-                        self._event_buffer.put(Event(element, key=key, time=time_as_datetime))
+                        event = Event(element, key=key, time=time_as_datetime)
+                        if create_event:
```
Comment (gtopper) [inline storey/sources.py]: Better move the work for creating the event into the if, to avoid unnecessary work.

File: storey/sources.py [hunk 3]
```diff
@@ -703,17 +718,28 @@ async def _run_loop(self):
...
                 if self._time_field:
                     time = body[self._time_field]
                 id = None
                 if self._id_field:
                     id = body[self._id_field]
-                event = Event(body, key=key, time=time, id=id)
-                await self._do_downstream(event)
+                if create_event:
+                    event = Event(body, key=key, time=time, id=id)
+                    await self._do_downstream(event)
```
Comment (gtopper) [inline storey/sources.py]: Better move the work for creating the event into the if, to avoid unnecessary work.

File: tests/test_flow.py [hunk 4]
```diff
@@ -2556,6 +2556,21 @@ def test_custom_string_time_input():
...
+
+    controller = build_flow([
+        DataframeSource(data, key_field=['first_name']),
+        Reduce([], append_and_return),
+    ]).run()
+    result = controller.await_termination()
+    expected = [{'first_name': 'moshe', 'some_data': 1}, {'first_name': 'katya', 'some_data': 3}]
+
+    assert result == expected, \
+        f'actual did not match expected. \n actual: {result} \n expected: {expected}'
```
Comment (gtopper) [inline tests/test_flow.py]: No need for this - pytest already prints this stuff. I know some tests (aggr...) do this, but none of the tests in this file do.

File: storey/sources.py (hunk 3 above)
Comment (katyakats) [reply storey/sources.py]: it;s in the if :)

Comment (gtopper) [reply storey/sources.py]: I meant the
```
                time = None
                if self._time_field:
                    time = body[self._time_field]
                id = None
                if self._id_field:
                    id = body[self._id_field]
```
stuff.

File: storey/sources.py [hunk 5]
```diff
@@ -619,20 +620,33 @@ def _blocking_io_loop(self):
...
+                                    create_event = False
                         if self._time_field:
                             time_field = self._time_field
                             if self._with_header and isinstance(time_field, str):
                                 time_field = field_name_to_index[time_field]
                             time_as_datetime = parsed_line[time_field]
                         else:
                             time_as_datetime = datetime.now()
-                        self._event_buffer.put(Event(element, key=key, time=time_as_datetime))
+                        if create_event:
```
Comment (gtopper) [inline storey/sources.py]: Here too, I meant the `time_field` stuff, that's not needed in case we don't create an event.

**Review by gtopper — APPROVED** [review]

**Pull request merged**